### PR TITLE
feat: Adding support for Expr.map_batches

### PR DESCRIPTION
**Pull request by DeaMariaLeon**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes # https://github.com/narwhals-dev/narwhals/issues/1226

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below.
I don't know if this is what was expected and if I should add the expression to dask as well.

### CODE REVIEW

**Comment by DeaMariaLeon** [comment]
Before I try to fix this for older versions, I would like to know if I took the right approach.

**Comment by MarcoGorelli** [comment]
hey, thanks for working on this! from an initial review, this looks really solid! will do a full review shortly

**Review by MarcoGorelli — COMMENTED** [review]
thanks for working on this!

I think the `output_names` part might not be quite right, as if the output names aren't known in advance (e.g. with `nw.all()`), then we get an empty dataframe back:
```python

data = {'a': [1,2,3], 'b': [4,5,6]}

print(nw.from_native(pd.DataFrame(data)).select(nw.all().map_batches(lambda x: x.to_numpy())).to_native())
print(nw.from_native(pl.DataFrame(data)).select(nw.all().map_batches(lambda x: x.to_numpy())).to_native())
```
outputs
```
Empty DataFrame
Columns: []
Index: []
shape: (3, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 1   ┆ 4   │
│ 2   ┆ 5   │
│ 3   ┆ 6   │
└─────┴─────┘

```

**Comment by DeaMariaLeon** [comment]
I fixed the issue with the names @MarcoGorelli. (I'm not sure if I have to ping you, no rush of course). 
Again, older versions are complaining, but maybe you can tell me about the rest? thanks.

File: narwhals/expr.py [hunk 1]
```diff
@@ -495,6 +495,71 @@ def std(self, *, ddof: int = 1) -> Self:
...
+            >>> df_pa = pa.table(data)
+
+            Let's define a dataframe-agnostic function:
+
+            >>> @nw.narwhalify
+            ... def func(df):
+            ...     return df.select(
+            ...         nw.col("a", "b")
+            ...         .map_batches(lambda s: s.to_numpy() + 1, return_dtype=nw.Float64)
+            ...         .sum()
```
[the remark this answers is not in the document]
Comment (DeaMariaLeon) [reply narwhals/expr.py]: Removed.

**Comment by DeaMariaLeon** [comment]
It seems that `map_batches` has a bug in old polars (0.20.3). Doing the same operation than the test that was failing (with just polars) there is the same error.

```
import polars as pl
data = {"a": [1, 2, 3], "b": [4, 5, 6], "z": [7.0, 8.0, 9.0]}
df_pl = pl.DataFrame(data)
print(pl.Float64())
print(df_pl.select(pl.all().map_batches(lambda s: s.to_numpy() + 1, return_dtype=pl.Float64())))
```

**Comment by MarcoGorelli** [comment]
thanks for updating! this is probably good to go, will do a full review soon-ish

**Comment by MarcoGorelli** [comment]
i just tried this out, and it looks like we're getting different outputs across different versions:
```
# ruff: noqa
from typing import Any
import pandas as pd
import polars as pl
import narwhals as nw
import pyarrow as pa
data = {"statistic": [1, 2, -3, 4]}
df_pd = pd.DataFrame(data)
df_pl = pl.DataFrame(data)
df_pa = pa.table(data)

@nw.narwhalify
def p_value(df: nw.DataFrame[Any]) -> nw.DataFrame[Any]:
    return df.select(
        nw.all().map_batches(lambda x: x.to_numpy())
    )

print(p_value(df_pl))
print(p_value(df_pd))
print(p_value(df_pa))
```

I think it might be OK to raise an error if we end up in `            if is_numpy_array(result[0]):` and `self._output_names` is `None`

The other thing I noticed is that this currently raises with the output from the docstring in https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.map_batches.html (i.e. `pl.all().map_batches(lambda x: x.to_numpy().argmax())`)

```
AttributeError: 'numpy.int64' object has no attribute '__native_namespace__'. Did you mean: '__array_namespace__'?
```

**Comment by DeaMariaLeon** [comment]
>i just tried this out, and it looks like we're getting different outputs across different versions:

Which versions, do you know? It's passing the old versions in the tests. And locally I'm testing with an old version too and it works.. You didn't test with an older version of my PR, did you? (🤞) 
```
shape: (1, 1)
┌─────────────┐
│ statistic   │
│ ---         │
│ list[i64]   │
╞═════════════╡
│ [1, 2, … 4] │
└─────────────┘
   statistic
0          1
1          2
2         -3
3          4
pyarrow.Table
statistic: int64
----
statistic: [[1,2,-3,4]]
```

For the second part of your comment, ("The other thing I noticed is... ") . I'll work on it.

**Comment by MarcoGorelli** [comment]
sorry i meant different packages - for polars it outpus one column, for pandas and pyarrow the output is empty

**Comment by DeaMariaLeon** [comment]
So, I did the same example you posted, with the normal/recent versions, and I still have the right answers. (Just like on my previous comment). I'm lost. 🫣

**Comment by MarcoGorelli** [comment]
sure, just to clarify, this is what I mean:
```python
In [5]: from typing import Any
   ...: import pandas as pd
   ...: import polars as pl
   ...: import narwhals as nw
   ...: import pyarrow as pa
   ...: data = {"statistic": [1, 2, -3, 4]}
   ...: df_pd = pd.DataFrame(data)
   ...: df_pl = pl.DataFrame(data)
   ...: df_pa = pa.table(data)
   ...: 
   ...: @nw.narwhalify
   ...: def func(df: nw.DataFrame[Any]) -> nw.DataFrame[Any]:
   ...:     return df.select(
   ...:         nw.all().map_batches(lambda x: x.to_numpy())
   ...:     )
   ...: 

In [6]: func(pd.DataFrame(data))  # empty output
Out[6]: 
Empty DataFrame
Columns: []
Index: []

In [7]: func(pl.DataFrame(data))  # one column
Out[7]: 
shape: (4, 1)
┌───────────┐
│ statistic │
│ ---       │
│ i64       │
╞═══════════╡
│ 1         │
│ 2         │
│ -3        │
│ 4         │
└───────────┘
```

**Comment by MarcoGorelli** [comment]
ah sorry my bad 🤦 i had indeed to updated my branch because of a conflict, so sorry

**Comment by MarcoGorelli** [comment]
after all these years i'm still making git mistakes

**Comment by DeaMariaLeon** [comment]
Hey, that's OK! No need to apologise.

>after all these years i'm still making git mistakes

We are only humans. 🤗

**Comment by DeaMariaLeon** [comment]
Fixed this part, it seems: 
>The other thing I noticed is that this currently raises with the output from the docstring in https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.map_batches.html (i.e. pl.all().map_batches(lambda x: x.to_numpy().argmax()))

File: narwhals/_arrow/expr.py [hunk 2]
```diff
@@ -393,11 +394,16 @@ def func(df: ArrowDataFrame) -> list[ArrowSeries]:
             input_series_list = self._call(df)
             output_names = [input_series.name for input_series in input_series_list]
             result = [function(series) for series in input_series_list]
-            if is_numpy_array(result[0]):
+
+            if is_numpy_array(result[0]) or is_numpy_generic(result[0]):
```
Comment (MarcoGorelli) [inline narwhals/_arrow/expr.py:398]: does it work to just use `numpy.isscalar`?

**Review by MarcoGorelli — COMMENTED** [review]
thanks for updating!

**Review by MarcoGorelli — APPROVED** [review]
should be good, thanks @DeaMariaLeon !

**Pull request merged**